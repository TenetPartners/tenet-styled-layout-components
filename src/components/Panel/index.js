import React from 'react';
import styled, { css } from 'styled-components';
import breakpoint from 'styled-components-breakpoint';

export default styled(({
  color,
  backgroundColor,
  imagePosition,
  grid,
  rows,
  maxWidth,
  ...rest
}) => <section {...rest} />)`
  ${({ maxWidth }) => maxWidth && css`
    ${breakpoint('max')`
      padding-left: calc(50vw - ${maxWidth / 16 / 2}em);
      padding-right: calc(50vw - ${maxWidth / 16 / 2}em);
    `}
  `}

  background-color: ${({ backgroundColor }) => backgroundColor};
  color: ${({ color }) => color};

  ${({ backgroundImage }) => backgroundImage && css`
    display: grid;

    > img {
      grid-column: 1 / -1;
      grid-row: 1 / -1;
      width: 100vw;
      max-width: none;

      ${({ maxWidth }) => maxWidth && css`
        ${breakpoint('max')`
          margin-left: calc((50vw - ${maxWidth / 16 / 2}em) * -1);
          margin-right: calc((50vw - ${maxWidth / 16 / 2}em) * -1);
        `}
      `}
    }

    > .content-container {
      grid-column: 1 / -1;
      grid-row: 1 / -1;
    }
  `}
`;
