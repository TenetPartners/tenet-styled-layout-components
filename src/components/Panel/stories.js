/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import Panel from '.';

const breakpoints = {
  mobile: 0,
  tablet: 737,
  desktop: 1025,
  max: 1200,
};

storiesOf('Panel', module)
  .add('max width of 1200', () => (
    <ThemeProvider theme={{ breakpoints }}>
      <Panel maxWidth={1200}>
        <p>hello</p>
        <p>world</p>
      </Panel>
    </ThemeProvider>
  ))
  .add('max width of 1200, background color blue', () => (
    <ThemeProvider theme={{ breakpoints }}>
      <Panel maxWidth={1200} backgroundColor="lightblue">
        <p>hello</p>
        <p>world</p>
      </Panel>
    </ThemeProvider>
  ))
  .add('max width of 1200, background color black, font color white', () => (
    <ThemeProvider theme={{ breakpoints }}>
      <Panel maxWidth={1200} backgroundColor="black" color="white">
        <p>hello</p>
        <p>world</p>
      </Panel>
    </ThemeProvider>
  ))
  .add('max width of 1200, background image', () => (
    <ThemeProvider theme={{ breakpoints }}>
      <Panel maxWidth={1200} backgroundImage>
        <img src="https://via.placeholder.com/1200x480" alt="placeholder" />
        <div className="content-container">
          <p>hello</p>
          <p>world</p>
        </div>
      </Panel>
    </ThemeProvider>
  ));
