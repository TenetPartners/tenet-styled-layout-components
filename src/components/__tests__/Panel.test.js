/* eslint-disable function-paren-newline */
import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import Panel from '../Panel';

const breakpoints = {
  mobile: 0,
  tablet: 737,
  desktop: 1025,
  max: 1200,
};

describe('Panel', () => {
  it('renders correctly for max width', () => {
    const tree = renderer.create(
      <ThemeProvider theme={{ breakpoints }}>
        <Panel maxWidth={1200}>
          <p>hello</p>
          <p>world</p>
        </Panel>
      </ThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('padding-left', 'calc(50vw - 37.5em)', {
      media: '(min-width:75em)',
    });
  });

  it('renders correctly for background color', () => {
    const tree = renderer.create(
      <ThemeProvider theme={{ breakpoints }}>
        <Panel backgroundColor="black">
          <p>hello</p>
          <p>world</p>
        </Panel>
      </ThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('background-color', 'black');
  });

  it('renders correctly for color', () => {
    const tree = renderer.create(
      <Panel color="blue">
        <p>hello</p>
        <p>world</p>
      </Panel>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('color', 'blue');
  });

  it('renders correctly for background image', () => {
    const tree = renderer.create(
      <ThemeProvider theme={{ breakpoints }}>
        <Panel maxWidth={1200} backgroundImage>
          <img src="https://via.placeholder.com/1200x480" alt="placeholder" />
          <div className="content-container">
            <p>hello</p>
            <p>world</p>
          </div>
        </Panel>
      </ThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('display', 'grid');
    expect(tree).toHaveStyleRule('grid-column', '1 / -1', {
      modifier: '> img',
    });
    expect(tree).toHaveStyleRule('grid-row', '1 / -1', {
      modifier: '> img',
    });
    expect(tree).toHaveStyleRule('width', '100vw', {
      modifier: '> img',
    });
    expect(tree).toHaveStyleRule('max-width', 'none', {
      modifier: '> img',
    });
    expect(tree).toHaveStyleRule('margin-left', 'calc((50vw - 37.5em) * -1)', {
      modifier: '> img',
      media: '(min-width:75em)',
    });
    expect(tree).toHaveStyleRule('margin-right', 'calc((50vw - 37.5em) * -1)', {
      modifier: '> img',
      media: '(min-width:75em)',
    });
    expect(tree).toHaveStyleRule('grid-column', '1 / -1', {
      modifier: '> .content-container',
    });
    expect(tree).toHaveStyleRule('grid-row', '1 / -1', {
      modifier: '> .content-container',
    });
  });
});
