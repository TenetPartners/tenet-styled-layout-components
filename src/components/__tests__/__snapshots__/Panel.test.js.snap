// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel renders correctly for background color 1`] = `
.c0 {
  background-color: black;
}

<section
  className="c0"
>
  <p>
    hello
  </p>
  <p>
    world
  </p>
</section>
`;

exports[`Panel renders correctly for background image 1`] = `
.c0 {
  display: grid;
}

.c0 > img {
  grid-column: 1 / -1;
  grid-row: 1 / -1;
  width: 100vw;
  max-width: none;
}

.c0 > .content-container {
  grid-column: 1 / -1;
  grid-row: 1 / -1;
}

@media (min-width:75em) {
  .c0 {
    padding-left: calc(50vw - 37.5em);
    padding-right: calc(50vw - 37.5em);
  }
}

@media (min-width:75em) {
  .c0 > img {
    margin-left: calc((50vw - 37.5em) * -1);
    margin-right: calc((50vw - 37.5em) * -1);
  }
}

<section
  backgroundImage={true}
  className="c0"
>
  <img
    alt="placeholder"
    src="https://via.placeholder.com/1200x480"
  />
  <div
    className="content-container"
  >
    <p>
      hello
    </p>
    <p>
      world
    </p>
  </div>
</section>
`;

exports[`Panel renders correctly for color 1`] = `
.c0 {
  color: blue;
}

<section
  className="c0"
>
  <p>
    hello
  </p>
  <p>
    world
  </p>
</section>
`;

exports[`Panel renders correctly for max width 1`] = `
@media (min-width:75em) {
  .c0 {
    padding-left: calc(50vw - 37.5em);
    padding-right: calc(50vw - 37.5em);
  }
}

<section
  className="c0"
>
  <p>
    hello
  </p>
  <p>
    world
  </p>
</section>
`;
